You're right to ask for disclosure and evidence. I used AI assistance to write this up
and didn't say so. That's on me. I'll also update the description to disclose it there and
drop the AI-style formatting.

I run several freqtrade bots on MariaDB and hit this after upgrading to 2026.4. That's
where this PR came from. I've gone through the final diff myself.

Root cause: MySQL and MariaDB have no native boolean type. `BOOLEAN` is an alias for
`TINYINT(1)`:

- MariaDB: https://mariadb.com/kb/en/boolean/
- MySQL: https://dev.mysql.com/doc/refman/8.4/en/numeric-type-syntax.html

`WalletHistory.bot_managed` is declared `Mapped[bool]`, so on these backends the column
is `TINYINT(1)`. With MariaDB + PyMySQL + pandas 3.0.3, `read_sql` reads it back as an
`int64` column, not `bool`. `results.loc[results["bot_managed"]]` in
`_rpc_get_historic_balance` then does label-based selection instead of boolean masking.

I overstated the impact in the original description: it doesn't always 500. `read_sql`
gives the frame a default `RangeIndex`, so the outcome depends on row count:

- `wallet_history` has one bot-managed row: `.loc` looks up label `1`, which the index
  doesn't contain, so it raises `KeyError` and the request 500s.
- 2+ rows: labels `0` and `1` both exist, so no error, but `.loc` selects rows by label
  instead of filtering, returning wrong/duplicated rows with HTTP 200.

So a fresh MariaDB bot 500s on this endpoint until `wallet_history` has a second row,
after which it silently returns the wrong balance history. `.astype(bool)` fixes both.
The route handler `api_get_wallet_history` calls `_rpc_get_historic_balance()` without
catching `KeyError`, which is why it surfaces as a 500.

Reproduction, using freqtrade's own model and the real `_rpc_get_historic_balance`
against a MariaDB server, no mocks:

```python
# pip install freqtrade pymysql        (freqtrade 2026.4)
from datetime import datetime
from freqtrade.persistence import init_db
from freqtrade.persistence.wallet_history import WalletHistory
from freqtrade.rpc.rpc import RPC

init_db("mysql+pymysql://user:pass@127.0.0.1/freqtrade_test")  # any MySQL/MariaDB schema

WalletHistory.session.add(WalletHistory(
    timestamp=datetime(2024, 1, 1), currency="USDT", quote_currency="USDT",
    balance=100.0, total_quote=100.0, bot_managed=True))
WalletHistory.session.commit()

RPC._rpc_get_historic_balance(None)   # what the /historic_balance route calls
```

<details>
<summary>Output (freqtrade 2026.4, pandas 3.0.3, MariaDB)</summary>

```
Traceback (most recent call last):
  File "repro.py", line 14, in <module>
    RPC._rpc_get_historic_balance(None)
  File ".../freqtrade/rpc/rpc.py", line 799, in _rpc_get_historic_balance
    results_filtered = results.loc[results["bot_managed"]]
  [... pandas .loc label indexing ...]
KeyError: 'None of [RangeIndex(start=1, stop=2, step=1)] are in the [index]'
```
</details>

I don't have the original 500s from my bots' own logs, since the containers were
redeployed when I applied a local patch, so the above is a minimal reproduction rather
than production log lines. It's freqtrade's own code and runs against any empty
MySQL/MariaDB schema.

`test_api_historic_balance` stays green because CI runs on SQLite, where the column is
a real `bool`. The test I added builds the `int64` column shape; it fails on `develop`
and passes with the fix. Happy to move the coercion to the query or model level if
you'd prefer.

--- a/freqtrade/rpc/rpc.py
+++ b/freqtrade/rpc/rpc.py
@@ -795,9 +795,7 @@ class RPC:
 
         results = results.rename({"timestamp": "date"}, axis=1)
         results.loc[:, "__date_ts"] = results.loc[:, "date"].dt.as_unit("ms").astype("int64")
-        # Exclude non-bot managed rows. Coerce to bool: read_sql may return the
-        # Boolean column as an integer dtype (e.g. MySQL/MariaDB), which would
-        # make .loc do label-based indexing instead of boolean masking.
+        # Exclude non-bot managed for now
         results_filtered = results.loc[results["bot_managed"].astype(bool)]
 
         results_final = (

--- a/tests/rpc/test_rpc_apiserver.py
+++ b/tests/rpc/test_rpc_apiserver.py
@@ -1472,13 +1472,13 @@ def test_api_historic_balance(botclient, mocker, ticker, fee, markets, is_short)
 
 
 def test_api_historic_balance_int_bot_managed(botclient, mocker):
-    """Regression: read_sql may return the wallet_history `bot_managed` Boolean
-    column as an integer dtype (e.g. MySQL/MariaDB TINYINT); the `.loc[...]`
-    filter must boolean-mask, not label-index."""
+    """
+    read_sql may return the wallet_history `bot_managed` column as an
+    integer (e.g. MySQL/MariaDB TINYINT)
+    """
     _, client = botclient
 
-    # Single bot-managed row: with an int64 `bot_managed`, label-based `.loc`
-    # indexing looks up label 1, which the RangeIndex lacks -> KeyError -> 500.
+    # Single row: with an int64 `bot_managed`
     one_row = pd.DataFrame(
         {
             "timestamp": pd.to_datetime(["2024-01-01"]),
@@ -1490,6 +1490,8 @@ def test_api_historic_balance_int_bot_managed(botclient, mocker):
     rc = client_get(client, f"{BASE_URI}/historic_balance")
     assert_response(rc, 200)
     assert rc.json()["length"] == 1
+    assert rc.json()["data"][0][0] == "2024-01-01T00:00:00"
+    assert rc.json()["data"][0][2] == 100.0
 
     # Mixed rows: the non-bot-managed row (bot_managed=0) must be excluded.
     two_rows = pd.DataFrame(
@@ -1503,6 +1505,8 @@ def test_api_historic_balance_int_bot_managed(botclient, mocker):
     rc = client_get(client, f"{BASE_URI}/historic_balance")
     assert_response(rc, 200)
     assert rc.json()["length"] == 1
+    assert rc.json()["data"][0][0] == "2024-01-02T00:00:00"
+    assert rc.json()["data"][0][2] == 200.0
 
 
 def test_api_performance(botclient, fee):

